File: webconsole/js/mqtt-client.js
// 连接选项
const options = {
    connectTimeout: 10000, // 超时时间
}
var mqttClient;
var sendCallback;
function startMqtt(url, topic, recvCallback, _sendCallback) {
	// 连接选项
	var options = {
		connectTimeout: 5000, // 超时时间
		username: 'admin',
		password: 'public',
	};

	mqttClient = mqtt.connect(url, options);

	mqttClient.on('reconnect', (error) => {
		append('正在重连:' + JSON.stringify(error));
	});

	mqttClient.on('error', (error) => {
		append('连接失败:' + JSON.stringify(error));
	});
	
	mqttClient.on('close', (e)=>{
		append("连接已关闭");
	});

	// 监听接收消息事件
	if(recvCallback){
		mqttClient.on('message', recvCallback);
	}else{
		mqttClient.on('message', (topic, message) => {
			append('收到来自'+topic+'的消息: '+message.toString());
		});
	}
	
	mqttClient.on('connect', (e) => {
		append('连接成功')
		var arr = new Array();
		if(isString(topic)){
			arr[0] = topic;
		}else{
			arr = topic;
		}
		for(var idx =0; idx < arr.length; idx++) {
			mqttClient.subscribe(topic, {qos: 0}, (error) => {
				if (!error) {
					append('订阅成功');
				} else {
					append('订阅失败: ' + JSON.stringify(error));
				}
			});
		}
	});
	if(_sendCallback){
		sendCallback = _sendCallback;
	}
	return mqttClient;
}

function stopMqtt(){
	if(mqttClient){
		mqttClient.end(true);
		mqttClient = null;
	}
}

// 取消订阅
function unsubscribe(topic){
    mqttClient.unubscribe(
        // topic, topic Array, topic Array-Onject
        topic
    );
}

// 发布消息
var default_send_qos = {qos: 0, retain: false};
function publish(topic, msg) {
    if (!mqttClient.connected) {
        console.log('客户端未连接');
        return;
    }
    if(!topic)
        topic = mqttDroneTopic;
    mqttClient.publish(topic, isString(msg) ? msg : JSON.stringify(msg), default_send_qos, sendCallback);
}

File: SshEndPoint/mqttclient.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#!/usr/bin/env python
# -*- coding: utf-8 -*-
'''
Created on 2019年5月5日

@author: russ
'''
import json
import shgutil
import paho.mqtt.client as mqtt
import uuid

global para_factory_local

def on_connect(mqttClient, obj, flags, rc):
    mqttClient.connected = True # 设置连接标志
    print(shgutil.current_time(), "云端" if mqttClient.insType == 1 else "本地",
          "mqtt连接已建立")
    if mqttClient.subscribeList and len(mqttClient.subscribeList) > 0:
        for topic in mqttClient.subscribeList:
            if topic:
                mqttClient.subscribe(topic, 0)
                print(shgutil.current_time(), "云端" if mqttClient.insType == 1 else "本地", "mqtt提交订阅", topic)
    # 发送上线通知
    online_para = para_factory_local.online()
    content = json.dumps(online_para, ensure_ascii=False)
    mqttClient.send(content)


def on_message(mqttClient, obj, msg):
    print(shgutil.current_time(), "云端" if mqttClient.insType == 1 else "本地", "mqtt接收: ",
          str(msg.payload, encoding='utf-8'))


def on_publish(mqttClient, obj, mid):
    print(shgutil.current_time(), "云端" if mqttClient.insType == 1 else "本地",
          "发送成功")


def on_subscribe(mqttClient, obj, mid, granted_qos):
    print(shgutil.current_time(), "云端" if mqttClient.insType == 1 else "本地",
          "订阅成功")


def on_log(mqttClient, obj, level, string):
    print(shgutil.current_time(), "云端" if mqttClient.insType == 1 else "本地",
          "mqtt日志", string)


def on_disconnect(mqttClient, userdata, rc):
    mqttClient.connected = False # 设置连接标志
    print(shgutil.current_time(), "云端" if mqttClient.insType == 1 else "本地",
          "mqtt连接断开!")


class MqttClient:
    def __init__(self, host, port, username, password, serverTopic, insType=1, subscribeList=[], para_factory = None):
        # insType 1=>与云端的连接; 0=>与本地的连接
        global para_factory_local
        self.host = host
        self.port = port
        self.username = username
        self.password = password
        self.serverTopic = serverTopic
        self.mqttClient = None
        self.insType = insType
        self.subscribeList = subscribeList
        self.para_factory = para_factory
        para_factory_local = para_factory
        
    def start(self, on_message_callback = None, willMsg = None, protocol="websockets"):
        self.mqttClient = mqtt.Client(client_id=str(uuid.uuid1()), transport=protocol)
        self.mqttClient.connected = False # 设置连接标志
        self.mqttClient.insType = self.insType
        self.mqttClient.subscribeList = self.subscribeList
        self.mqttClient.ws_set_options()
        if self.username and self.password:
            self.mqttClient.username_pw_set(self.username, self.password)
        self.mqttClient.reconnect_delay_set(1, 5)
        if on_message_callback:
            self.mqttClient.on_message = on_message_callback
        self.mqttClient.on_connect = on_connect
        self.mqttClient.on_publish = on_publish
        self.mqttClient.on_subscribe = on_subscribe
        self.mqttClient.on_disconnect = on_disconnect
        #mqttClient.on_log = on_log
        self.mqttClient.disable_logger()
        if willMsg:
            self.mqttClient.will_set(self.serverTopic, willMsg)
        self.mqttClient.connect(self.host, self.port, 60)
        self.mqttClient.loop_start()

    def stop(self):
        self.mqttClient.disconnect()

    def reconnect(self):
        if not self.mqttClient.connected:
            self.mqttClient.connect(self.host, self.port, 60)

    def subscribe(self, topic, qos=0):
        self.mqttClient.subscribe(topic, qos)

    def unsubscribe(self, topic):
        self.mqttClient.unsubscribe(topic)

    def send(self, content, topic=None, qos=1, retain=False):
        if not content:
            return None
        #if not self.mqttClient.connected:
        #    print(shgutil.current_time(), "mqtt无发送, 发送任务取消")
        try:
            if topic is None:
                self.mqttClient.publish(self.serverTopic, content, qos, retain)
            else:
                self.mqttClient.publish(topic, content, qos, retain)
            return True
        except Exception as err:
            print(shgutil.current_time(), "mqtt发送异常1", err)
        return False
    
    def connected(self):
        return self.mqttClient.connected if self.mqttClient else False
    